File: among_them/modulabot/nim_perception/src/sprite_match.nim
## Sprite-matching kernels for the modulabot FFI.
##
## Exported symbols (Phase 1):
##
## - ``mb_match_actor_sprite_all`` — vectorised all-anchors match.
##   Semantics must be identical to
##   :func:`modulabot.sprite_match.matches_actor_sprite` run at every
##   valid anchor, under a caller-supplied miss / stable / tint budget.
## - ``mb_actor_color_index_all`` — per-anchor dominant-tint colour
##   index, matching :func:`modulabot.sprite_match.crewmate_color_index`.
##
## Palette constants are hardcoded here to match
## :mod:`modulabot.data` (``TintColor = 3``, ``ShadeTintColor = 9``,
## ``TransparentIndex = 255``, ``PLAYER_COLORS`` / ``SHADOW_MAP``
## identical to ``sim.nim``'s ``PlayerColors`` / ``ShadowMap``). If
## either side ever rewrites the palette, bump
## :data:`ModulabotPerceptionAbiVersion` in ``../lib.nim`` so the
## Python loader refuses the stale library.
##
## Data ownership: every buffer is caller-allocated. We read ``frame``
## and ``sprite`` as ``ptr UncheckedArray[uint8]`` and write through
## ``out_mask`` / ``out_indices`` in place. No allocation, no captured
## references — safe to call from any thread once the library is
## loaded (the Python GIL serialises the ctypes call anyway).

const
  ScreenWidth* = 128
  ScreenHeight* = 128
  TransparentIndex* = 255'u8
  TintColor* = 3'u8
  ShadeTintColor* = 9'u8
  PlayerColorCount* = 16
  PaletteMask* = 0x0F'u8

# Must match ``modulabot.data.PLAYER_COLORS`` (and ``sim.nim``'s
# ``PlayerColors``). Index into the array is the tracked player-colour
# slot; value is the palette index for that slot's lit tint.
const PlayerColors*: array[PlayerColorCount, uint8] = [
  3'u8, 7, 8, 14, 4, 11, 13, 15, 1, 2, 5, 6, 9, 10, 12, 0
]

# ``data.SHADOW_MAP``. Palette index → shadowed palette index. Exactly
# matches ``sim.nim``'s ``ShadowMap`` array.
const ShadowMap*: array[16, uint8] = [
  0'u8, 12, 9, 5, 5, 0, 5, 5, 5, 12, 9, 9, 0, 12, 12, 9
]

# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------

proc isPlayerBodyColor(c: uint8): bool {.inline.} =
  ## True iff ``c`` is a plausible player-body colour (lit tint *or*
  ## its shadowed variant). Mirrors
  ## :func:`modulabot.sprite_match.player_body_color`.
  ##
  ## Implemented as a linear scan rather than a 256-entry LUT because
  ## (a) the set is small (32 colours after dedup), (b) JIT-style
  ## branch prediction on the hot path is fine at this scale, and
  ## (c) a LUT would need to live in a ``let`` block with init, which
  ## is friction we don't need.
  for pc in PlayerColors:
    if c == pc:
      return true
    if c == ShadowMap[pc and PaletteMask]:
      return true
  return false

proc playerColorIndexOf(c: uint8): int {.inline.} =
  ## Return the tracked-player-colour slot for a *lit* (non-shadowed)
  ## palette index, or ``-1``. Matches
  ## :func:`modulabot.sprite_match.player_color_index`.
  for i, pc in PlayerColors:
    if c == pc:
      return i
  return -1

# ---------------------------------------------------------------------------
# mb_match_actor_sprite_all
# ---------------------------------------------------------------------------

proc mb_match_actor_sprite_all*(
    frame: ptr UncheckedArray[uint8],    # (SH*SW,) uint8 row-major
    sprite: ptr UncheckedArray[uint8],   # (sh*sw,) uint8 row-major
    sh: cint,
    sw: cint,
    flip_h: cint,
    max_misses: cint,
    min_stable: cint,
    min_tint: cint,
    out_mask: ptr UncheckedArray[uint8], # (max_y*max_x,) uint8 0/1
) {.exportc, dynlib.} =
  ## Vectorised all-anchors sprite match.
  ##
  ## Writes ``out_mask[ay * max_x + ax] = 1`` when the sprite is an
  ## acceptable match at anchor ``(ax, ay)`` under the budgets,
  ## otherwise ``0``.
  ##
  ## Semantics (per anchor):
  ##
  ## - **Stable** sprite pixels (neither ``TintColor`` nor
  ##   ``ShadeTintColor``, and not ``TransparentIndex``) must match
  ##   the frame colour exactly.
  ## - **Tint** sprite pixels (either ``TintColor`` or
  ##   ``ShadeTintColor``) match any plausible player body colour
  ##   (lit or shadowed).
  ## - Anchor is accepted iff cumulative misses ≤ ``max_misses`` AND
  ##   matched-stable ≥ ``min_stable`` AND matched-tint ≥ ``min_tint``.
  ##
  ## Performance: the outer loop is over anchors (cheap index math),
  ## the inner loop is over sprite pixels (bounded by sh*sw). With
  ## early-out on the miss budget the average anchor cost is far below
  ## the worst case because most anchors reject after a handful of
  ## miss hits.
  let
    maxY = int(ScreenHeight) - int(sh) + 1
    maxX = int(ScreenWidth) - int(sw) + 1
  if maxY <= 0 or maxX <= 0:
    return

  # Pre-count stable/tint *pixels* (not matches); bail early if the
  # sprite simply can't clear the floor because it doesn't have
  # enough of each. Mirrors the Python scalar path's final guard but
  # lets us skip the whole anchor sweep in the degenerate case.
  var totalStable = 0
  var totalTint = 0
  for sy in 0 ..< int(sh):
    for sx in 0 ..< int(sw):
      let c = sprite[sy * int(sw) + sx]
      if c == TransparentIndex: continue
      if c == TintColor or c == ShadeTintColor:
        inc totalTint
      else:
        inc totalStable
  if totalStable < int(min_stable) or totalTint < int(min_tint):
    # Zero the output buffer and return. Not strictly required — the
    # Python caller zeroes before calling — but belt-and-braces.
    for i in 0 ..< maxY * maxX:
      out_mask[i] = 0'u8
    return

  let flip = flip_h != 0

  for ay in 0 ..< maxY:
    for ax in 0 ..< maxX:
      var misses = 0
      var matchedStable = 0
      var matchedTint = 0
      var reject = false
      for sy in 0 ..< int(sh):
        if reject: break
        for sx in 0 ..< int(sw):
          let srcX = if flip: int(sw) - 1 - sx else: sx
          let c = sprite[sy * int(sw) + srcX]
          if c == TransparentIndex: continue
          let fy = ay + sy
          let fx = ax + sx
          # Anchor ranges above guarantee fy / fx in-bounds.
          let fc = frame[fy * int(ScreenWidth) + fx]
          if c == TintColor or c == ShadeTintColor:
            if isPlayerBodyColor(fc):
              inc matchedTint
            else:
              inc misses
          else:
            if fc == c:
              inc matchedStable
            else:
              inc misses
          if misses > int(max_misses):
            reject = true
            break
      if reject or matchedStable < int(min_stable) or matchedTint < int(min_tint):
        out_mask[ay * maxX + ax] = 0'u8
      else:
        out_mask[ay * maxX + ax] = 1'u8

# ---------------------------------------------------------------------------
# mb_actor_color_index_all
# ---------------------------------------------------------------------------

proc mb_actor_color_index_all*(
    frame: ptr UncheckedArray[uint8],
    sprite: ptr UncheckedArray[uint8],
    sh: cint,
    sw: cint,
    flip_h: cint,
    out_indices: ptr UncheckedArray[int8], # (max_y*max_x,) int8
) {.exportc, dynlib.} =
  ## Per-anchor dominant-tint colour index.
  ##
  ## For every anchor ``(ax, ay)``, count how many of the sprite's
  ## ``TintColor`` pixels match each tracked player colour in the
  ## frame (lit tint only — shadowed palette indices do not vote).
  ## Write the argmax index into ``out_indices`` or ``-1`` if no
  ## tint pixel voted.
  ##
  ## Ties are broken by lowest index, matching Python's
  ## ``np.argmax``.
  let
    maxY = int(ScreenHeight) - int(sh) + 1
    maxX = int(ScreenWidth) - int(sw) + 1
  if maxY <= 0 or maxX <= 0:
    return

  let flip = flip_h != 0
  var counts: array[PlayerColorCount, int32]

  for ay in 0 ..< maxY:
    for ax in 0 ..< maxX:
      for i in 0 ..< PlayerColorCount:
        counts[i] = 0
      for sy in 0 ..< int(sh):
        for sx in 0 ..< int(sw):
          let srcX = if flip: int(sw) - 1 - sx else: sx
          let c = sprite[sy * int(sw) + srcX]
          if c != TintColor: continue
          let fy = ay + sy
          let fx = ax + sx
          let fc = frame[fy * int(ScreenWidth) + fx]
          let idx = playerColorIndexOf(fc)
          if idx >= 0:
            inc counts[idx]
      var best = 0
      var bestVotes = counts[0]
      for i in 1 ..< PlayerColorCount:
        if counts[i] > bestVotes:
          bestVotes = counts[i]
          best = i
      if bestVotes > 0:
        out_indices[ay * maxX + ax] = int8(best)
      else:
        out_indices[ay * maxX + ax] = -1'i8

File: players/among_them/coborg/perception/sprite_match.py
"""Sprite-matching kernels. Port of
``users/james/personal_cogs/among_them/common/perception_kernels/sprite_match.nim``.

Two public procs, names mirrored 1:1 to the Nim symbols (snake_cased) so the
parity rig can assert symbol-by-symbol equality:

- :func:`match_actor_sprite_all` -- all-anchors sprite match under
  stable/tint/transparent/budget semantics.
- :func:`actor_color_index_all` -- per-anchor dominant-tint player-color
  index (or -1 if no tint pixel voted).

Numpy-first per PLAN section 5.3. Both kernels use ``sliding_window_view``
to materialise per-anchor patches, then broadcast over the sprite mask.
Numba is intentionally left off and is only promoted if measurement under
``tests/test_sprite_match.py`` shows the perf budget breached.
"""

from __future__ import annotations

import numpy as np
from numpy.lib.stride_tricks import sliding_window_view

from .data.palette import (
    SHADE_TINT_COLOR,
    TINT_COLOR,
    TRANSPARENT_INDEX,
)
from .frame import SCREEN_HEIGHT, SCREEN_WIDTH

# Player color slots -> lit palette index. Must match
# ``sim.nim``'s ``PlayerColors`` and
# ``common/perception_kernels/sprite_match.nim``'s ``PlayerColors``. The
# 16 slots cover every PICO-8 palette entry exactly once (a permutation),
# which has implications for ``actor_color_index_all`` -- see below.
PLAYER_COLORS: np.ndarray = np.array(
    [3, 7, 8, 14, 4, 11, 13, 15, 1, 2, 5, 6, 9, 10, 12, 0], dtype=np.uint8
)
PLAYER_COLORS.flags.writeable = False

# Palette index -> shadowed variant. Mirrors ``sim.nim``'s ``ShadowMap``.
SHADOW_MAP: np.ndarray = np.array(
    [0, 12, 9, 5, 5, 0, 5, 5, 5, 12, 9, 9, 0, 12, 12, 9], dtype=np.uint8
)
SHADOW_MAP.flags.writeable = False


def _build_player_body_lut() -> np.ndarray:
    """256-entry bool LUT: index ``c`` is True iff ``c`` is a plausible
    player-body palette index (lit color OR its shadowed variant).

    Mirrors ``isPlayerBodyColor`` in the Nim kernel. The Nim version does a
    linear scan over PlayerColors per pixel; numpy prefers a precomputed LUT
    for SIMD-friendly indexing.
    """
    lut = np.zeros(256, dtype=bool)
    for pc in PLAYER_COLORS:
        lut[int(pc)] = True
        lut[int(SHADOW_MAP[int(pc) & 0x0F])] = True
    return lut


PLAYER_BODY_LUT: np.ndarray = _build_player_body_lut()
PLAYER_BODY_LUT.flags.writeable = False


def _build_palette_to_slot() -> np.ndarray:
    """256-entry LUT: palette index -> player-color slot.

    ``PLAYER_COLORS`` maps slot -> palette index; this is the inverse on
    the lit-color path. Used by :func:`actor_color_index_all` to convert
    per-tint frame samples into slot indices in one shot before the
    per-anchor histogram. Palette indices outside the 0..15 range map to
    a sentinel slot (``PLAYER_COLORS.size`` == 16) that drops out of the
    argmax via ``minlength``.
    """
    lut = np.full(256, PLAYER_COLORS.size, dtype=np.int64)
    for slot, c in enumerate(PLAYER_COLORS):
        lut[int(c)] = slot
    return lut


_PALETTE_TO_SLOT: np.ndarray = _build_palette_to_slot()
_PALETTE_TO_SLOT.flags.writeable = False


def match_actor_sprite_all(
    frame: np.ndarray,
    sprite: np.ndarray,
    flip_h: bool,
    max_misses: int,
    min_stable: int,
    min_tint: int,
) -> np.ndarray:
    """Vectorised all-anchors sprite match.

    Returns a ``(max_y, max_x)`` ``uint8`` 0/1 mask where
    ``max_y = SCREEN_HEIGHT - sh + 1`` and similarly for x. ``out[ay, ax] == 1``
    iff sprite anchored at ``(ay, ax)`` clears the budgets.

    Semantics (per anchor, mirroring the upstream Nim kernel):

    - **Stable** sprite pixels (not transparent, not tint, not shade-tint):
      must match the frame palette index exactly.
    - **Tint** sprite pixels (tint OR shade-tint): match if the frame pixel
      is any plausible player-body color (lit OR shadowed).
    - **Transparent** sprite pixels are ignored.
    - Anchor accepts iff ``misses <= max_misses`` AND
      ``matched_stable >= min_stable`` AND ``matched_tint >= min_tint``.

    Pre-flight short-circuit: if the sprite simply doesn't have enough
    stable / tint pixels to clear the floor, the whole sweep is skipped and
    the output is all-zero (matching the Nim early-out).
    """
    frame = np.asarray(frame, dtype=np.uint8)
    sprite = np.asarray(sprite, dtype=np.uint8)
    if flip_h:
        sprite = np.ascontiguousarray(sprite[:, ::-1])

    sh, sw = sprite.shape
    max_y = SCREEN_HEIGHT - sh + 1
    max_x = SCREEN_WIDTH - sw + 1
    if max_y <= 0 or max_x <= 0:
        return np.zeros((max(0, max_y), max(0, max_x)), dtype=np.uint8)

    transparent = sprite == TRANSPARENT_INDEX
    tint = (sprite == TINT_COLOR) | (sprite == SHADE_TINT_COLOR)
    stable = ~transparent & ~tint
    stable_positions = np.argwhere(stable)  # (n_stable, 2)
    tint_positions = np.argwhere(tint)      # (n_tint, 2)
    total_stable = stable_positions.shape[0]
    total_tint = tint_positions.shape[0]
    if total_stable < min_stable or total_tint < min_tint:
        return np.zeros((max_y, max_x), dtype=np.uint8)

    # (max_y, max_x, sh, sw) view onto frame; no copy.
    windows = sliding_window_view(frame, (sh, sw))

    # Extract per-anchor frame patches just at the sprite positions we care
    # about. n_stable + n_tint is typically ~60-80 out of sh*sw=144; doing
    # full-window comparisons here cost ~2x in benchmarks because the bool
    # AND with the sprite mask kept allocating (max_y, max_x, sh, sw) bool
    # buffers. This formulation hits only the elements we actually consume.
    sprite_stable_vals = sprite[stable_positions[:, 0], stable_positions[:, 1]]
    windows_stable = windows[:, :, stable_positions[:, 0], stable_positions[:, 1]]
    matched_stable = (windows_stable == sprite_stable_vals).sum(axis=-1)

    windows_tint = windows[:, :, tint_positions[:, 0], tint_positions[:, 1]]
    matched_tint = PLAYER_BODY_LUT[windows_tint].sum(axis=-1)
    misses = (total_stable - matched_stable) + (total_tint - matched_tint)

    accept = (
        (misses <= max_misses)
        & (matched_stable >= min_stable)
        & (matched_tint >= min_tint)
    )
    return accept.astype(np.uint8)


def actor_color_index_all(
    frame: np.ndarray,
    sprite: np.ndarray,
    flip_h: bool,
) -> np.ndarray:
    """Per-anchor dominant-tint color index.

    Returns a ``(max_y, max_x)`` ``int8`` array. ``out[ay, ax]`` is the
    index into ``PLAYER_COLORS`` that received the most tint-pixel votes
    at that anchor, or ``-1`` if no tint pixel voted.

    Mirrors ``mb_actor_color_index_all``. Important asymmetry with
    :func:`match_actor_sprite_all`: only sprite pixels equal to
    ``TINT_COLOR`` (the lit tint, palette 3) trigger color votes;
    ``SHADE_TINT_COLOR`` (palette 9) does **not** vote.

    Frame pixels also only count if they're a lit player color
    (``PLAYER_COLORS[i]``). Shadowed variants do not vote -- that's the
    asymmetry on the *frame* side that matches the upstream Nim.

    Ties are broken by lowest index, matching ``np.argmax`` default.
    """
    frame = np.asarray(frame, dtype=np.uint8)
    sprite = np.asarray(sprite, dtype=np.uint8)
    if flip_h:
        sprite = np.ascontiguousarray(sprite[:, ::-1])

    sh, sw = sprite.shape
    max_y = SCREEN_HEIGHT - sh + 1
    max_x = SCREEN_WIDTH - sw + 1
    if max_y <= 0 or max_x <= 0:
        return np.zeros((max(0, max_y), max(0, max_x)), dtype=np.int8)

    tint_positions = np.argwhere(sprite == TINT_COLOR)  # (n_tint, 2)
    if tint_positions.shape[0] == 0:
        return np.full((max_y, max_x), -1, dtype=np.int8)

    windows = sliding_window_view(frame, (sh, sw))
    # Pull just the tint-pixel frame samples per anchor: shape
    # (max_y, max_x, n_tint). n_tint ~ 30 vs sh*sw=144 for the player
    # sprite, so this is much cheaper than running 16 full-window compares.
    tint_frame = windows[:, :, tint_positions[:, 0], tint_positions[:, 1]]

    # Map every sampled palette index to its lit player-color slot in a
    # single LUT pass, then build the per-anchor histogram via bincount
    # on a flattened (anchor_idx, slot) composite index. This avoids 16
    # boolean comparisons / sums on the (max_y, max_x, n_tint) array and
    # is the meaningful perf win for the kernel.
    n_anchors = max_y * max_x
    n_slots = PLAYER_COLORS.size  # 16
    slots = _PALETTE_TO_SLOT[tint_frame]  # (max_y, max_x, n_tint), int64 in [0, 16]
    anchor_offsets = np.arange(n_anchors, dtype=np.int64).reshape(max_y, max_x, 1)
    # (n_slots + 1) to give the out-of-range sentinel a parking slot.
    flat_idx = (anchor_offsets * (n_slots + 1) + slots).ravel()
    counts_flat = np.bincount(flat_idx, minlength=n_anchors * (n_slots + 1))
    counts = counts_flat.reshape(max_y, max_x, n_slots + 1)[..., :n_slots]

    best = counts.argmax(axis=-1).astype(np.int8)
    no_vote = counts.sum(axis=-1) == 0
    best[no_vote] = -1
    return best
